Issue #6893: Generate Xpath Suppresion on escape xml symbols

AkMo3 opened the pull request:
## Generate Xpath Suppresion on escape xml symbols

Closes: #6893  

#### Test File
```
$ cat Test.java
class A {

  String quoteChar = "\"escaped\"" +
  "2";

  String lessChar = "<escaped" +
  "3";

  String ampersandChar = "&escaped" +
  "4";

  String greaterChar = ">escaped" +
  "5";

  String newLineChar = "escaped\n" +
  "6";

  String specialChar = "escaped\r" +
  "6";

  String aposChar = "'escaped'" +
  "7";
}
```

#### Configuration File
```
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.2//EN" "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="OperatorWrapCheck"/>
    </module>
</module>
```

#### Result of checkstyle with above config file
```
$ java -jar java -jar checkstyle-8.41-SNAPSHOT-all.jar -c .\config.xml .\Test.java
Starting audit...
[ERROR] H:\CheckstyleCheck\.\Test.java:3:36: '+' should be on a new line. [OperatorWrap]
[ERROR] H:\CheckstyleCheck\.\Test.java:6:32: '+' should be on a new line. [OperatorWrap]
[ERROR] H:\CheckstyleCheck\.\Test.java:9:37: '+' should be on a new line. [OperatorWrap]
[ERROR] H:\CheckstyleCheck\.\Test.java:12:35: '+' should be on a new line. [OperatorWrap]
[ERROR] H:\CheckstyleCheck\.\Test.java:15:36: '+' should be on a new line. [OperatorWrap]
[ERROR] H:\CheckstyleCheck\.\Test.java:18:36: '+' should be on a new line. [OperatorWrap]
[ERROR] H:\CheckstyleCheck\.\Test.java:21:33: '+' should be on a new line. [OperatorWrap]
Audit done.
Checkstyle ends with 7 errors.
```

#### Suppressions generated by checkstyle after doing changes with this config
```
$ java -jar checkstyle-8.41-SNAPSHOT-all.jar -g -c .\config.xml .\Test.java            
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suppressions PUBLIC
    "-//Checkstyle//DTD SuppressionXpathFilter Experimental Configuration 1.2//EN"
    "https://checkstyle.org/dtds/suppressions_1_2_xpath_experimental.dtd">
<suppressions>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='quoteChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='\&quot;escaped\&quot;']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='lessChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='&lt;escaped']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='ampersandChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='&amp;escaped']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='greaterChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='&gt;escaped']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='newLineChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='escaped\n']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='specialChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='escaped\r']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='aposChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='&apos;&apos;escaped&apos;&apos;']]"/>
</suppressions>
Checkstyle ends with 7 errors.
```
#### Result of checkstyle after attaching suppressions to config
``` 
$ java -jar checkstyle-8.41-SNAPSHOT-all.jar -c .\suppscon.xml .\Test.java
Starting audit...
Audit done.
```

### Results with  previous 8.40 build

#### Generated Suppressions 
```
$ java -jar H:\checkstyle-8.40-all.jar -g -c .\config.xml .\Test.java
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suppressions PUBLIC
    "-//Checkstyle//DTD SuppressionXpathFilter Experimental Configuration 1.2//EN"
    "https://checkstyle.org/dtds/suppressions_1_2_xpath_experimental.dtd">
<suppressions>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='quoteChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='\"escaped\"']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='lessChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='<escaped']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='ampersandChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='&escaped']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='greaterChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='>escaped']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='newLineChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='escaped\n']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='specialChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='escaped\r']]"/>
<suppress-xpath
       files="Test.java"
       checks="OperatorWrapCheck"
       query="/CLASS_DEF[./IDENT[@text='A']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='aposChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text=''escaped'']]"/>
</suppressions>
Checkstyle ends with 7 errors.
```
#### On using these supprressions
```
$ java -jar H:\checkstyle-8.39-all.jar -c .\suppscon.xml .\Test.java
com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module TreeWalker - cannot initialize module SuppressionXpathFilter - Unable to parse supp.xml - Element type "suppress-xpath" must be followed by either attribute specifications, ">" or "/>".
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:482)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:201)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:404)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:331)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module SuppressionXpathFilter - Unable to parse supp.xml - Element type "suppress-xpath" must be followed by either attribute specifications, ">" or "/>".
        at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:125)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:201)
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:477)
        ... 5 more
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to parse supp.xml - Element type "suppress-xpath" must be followed by either attribute specifications, ">" or "/>".
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:292)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:267)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:253)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionXpathFilter.finishLocalSetup(SuppressionXpathFilter.java:591)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:197)
        at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:120)
        ... 7 more
Caused by: org.xml.sax.SAXParseException; systemId: file:/H:/CheckstyleCheck/supp.xml; lineNumber: 9; columnNumber: 139; Element type "suppress-xpath" must be followed by either attribute specifications, ">" or "/>".
        at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:204)
        at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:178)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:400)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1471)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.seekCloseOfStartTag(XMLDocumentFragmentScannerImpl.java:1442)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanStartElement(XMLDocumentFragmentScannerImpl.java:1371)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2725)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:541)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:888)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:824)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
        at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:635)
        at com.puppycrawl.tools.checkstyle.XmlLoader.parseInputSource(XmlLoader.java:86)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:283)
        ... 12 more
Checkstyle ends with 1 errors.
```

AkMo3 commented:
>We then need another test that shows that an xpath being generated from a checkstyle violation with special characters, similar to the above reported issue, can be suppressed with the xpath string being generated and that there are no other issues. This is similar to our current regression tests at suppressionxpathfilter in the IT folder.

@rnveach Can you please explain the above line from your this [comment](https://github.com/checkstyle/checkstyle/issues/6893#issuecomment-541418449) . Also, please take a look above.

rnveach commented:
See the tests in https://github.com/checkstyle/checkstyle/tree/master/src/it/java/org/checkstyle/suppressionxpathfilter .

strkkk commented:
@AkMo3 I do not clearly understand what the problem is.

I see that problem seems to be fixed in issue.
And there is escape for apos
https://github.com/checkstyle/checkstyle/pull/9247/files#diff-f3c2babc8351808dde2410e32d60134b26f436d07fa26588fa41fbf951e9af53R372-R374

Can you explain what is the issue?

AkMo3 commented:
@strkkk Using the default escape character for Xpath as used in earlier used in [XMLLogger](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java#L248). 
The issue arising is that for a String like `String giveExcep = "'aposonbothside'";` the Xpath that will be created for suppression will contain this identifier as `[@text = '&apos;aposonbothside&apos;']` , 
which while being decode by converts to `[@text = ''aposonbothside'']` . 
Here this double occurrence of `'` is the root of the problem.
Rest it works fine. I have also added regression tests.

strkkk commented:
> while being decoded converts to [@text = ''aposonbothside''] .

Where this decoding happens? If it is some 3rd party lib, in my vision, it should be able to handle such cases.

AkMo3 commented:
I also tried to find that earlier but was not able to find that. But I am guessing that the conversion is taking place at this [line](https://github.com/checkstyle/checkstyle/blob/32cb8fc7aea1595c175c543a370b4c06fb41280d/src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java#L465) of ElementNode.java file. It is being done by Saxon 10.3 and module is net.sf.saxon.tree.util.Navigator.

AkMo3 commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,62 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+     * @param chr character from parent method encode() .
+     * @param string StringBuilder from parent method encode().
+     */
+    private static void fixEscape(char chr, StringBuilder string) {
```
Please suggest I fine method name that passes the Questionable name string check of team city. I tried a bunch of names but neither of them get past it.

strkkk replied:
correctEscapedChar()

AkMo3 reviewed (commented):
Please suggest a name that can be used in place method fixEscape.

AkMo3 commented:
GitHub, rebase

AkMo3 commented:
GitHub, rebase

AkMo3 commented:
GitHub, rebase

AkMo3 commented:
@strkkk Found this line in Saxon documentation **From Saxon 7.1, string delimiters can be doubled within the string to represent the delimiter itself: for example <xsl:value-of select='"He said, ""Go!"""'/>** . So just changed the encoding and it works perfectly fine now.

AkMo3 commented:
Should I add regression test?

strkkk commented:
>Should I add regression test?

@AkMo3 I should do everything that makes CI happy :)
Now there many jobs that failed, may be it is because of lack of tests

AkMo3 commented:
> 
> 
> > Should I add regression test?
> 
> @AkMo3 I should do everything that makes CI happy :)
> Now there many jobs that failed, may be it is because of lack of tests

Sorry, I just pushed without perfectly running mvn verify again after fixing errors. I am just waiting to finish mvn verify run smoothly then push.

AkMo3 commented:
@strkkk Sir, please suggest another name for that function. correctEscapedChar() also is not being accepted.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,62 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+     * @param chr character from parent method encode() .
+     * @param string StringBuilder from parent method encode().
+     */
+    private static void generateControlledString(char chr, StringBuilder string) {
```
just `encode` can be ok I guess

AkMo3 replied:
`encode` is already the main function

strkkk replied:
yeah, but signature is different so it will be overloaded method

AkMo3 replied:
Yup, got that. Sir, I had changed it to `controlTrueCharacter()` to give a better sense as `encode` might confuse later if some other contributors try to fix something. Will it be ok?

strkkk commented:
@AkMo3 I added comment.

Please request a review after CI is green

AkMo3 commented:
> 
> 
> @AkMo3 I added comment.
> 
> Please request a review after CI is green

Got it.

AkMo3 commented:
@strkkk Please reviews. I had updated the PR description covering all cases.

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -48,6 +50,26 @@
     private static FileText fileText;
 
+    private final CloseAndFlushTestByteArrayOutputStream outStream =
+            new CloseAndFlushTestByteArrayOutputStream();
```
what this stream is for?

AkMo3 replied:
This code is redundant, I will remove it. I added it while testing something. Thanks for pointing out.

AkMo3 replied:
Done

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -48,6 +50,26 @@
+            {"\u0080", "#x80;"},
+            {"\n", "&#10;"},
+            {"\r", ""},
+    };
```
please change to map and make local variable

AkMo3 replied:
Ok.

AkMo3 replied:
Done

strkkk replied:
@AkMo3 it is changed to map, but it is not a local variable but field. This map is not needed anywhere else but your test method, so please move it to your test

AkMo3 replied:
Done.

strkkk commented on src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionMatchXpathTest.java:
```diff
@@ -87,4 +88,154 @@ public void testTwo() throws Exception {
     }
+
+    @Test
+    public void testEncodedOne() throws Exception {
```
Please also add test cases to SuppressionXpathFilterTest with escape sequences in file

AkMo3 replied:
On it.

AkMo3 replied:
Please see my comment. I am having some issues related to this teset.

AkMo3 replied:
Done

strkkk commented:
@AkMo3 there is one more thing I forgot about - there was same issue with escapes in char literals, e.g. `char a = '&'`
Please extend test cases with this case (and other chars) as well.

Note - to produce violation, you can use IllegalTokenCheck with tokens = CHAR_LITERAL

AkMo3 commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,61 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+     * Controls true escape characters. True escape characters need '&' before but it also
+     * requires XML 1.1 until https://github.com/checkstyle/checkstyle/issues/5168.
+     *
+     * @param chr character from parent method encode() .
```
On it.

AkMo3 replied:
Done

AkMo3 commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,61 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+    }
+
+    /**
+     * Escape &lt;, &gt; &amp; &#39; and &quot; as their entities.
```
On it.

AkMo3 replied:
Done

AkMo3 commented:
> 
> 
> @AkMo3 there is one more thing I forgot about - there was same issue with escapes in char literals, e.g. `char a = '&'`
> Please extend test cases with this case (and other chars) as well.
> 
> Note - to produce violation, you can use IllegalTokenCheck with tokens = CHAR_LITERAL

Ok, I would look into this.

AkMo3 commented on src/it/resources/org/checkstyle/suppressionxpathfilter/matchxpath/SuppressionXpathRegressionMatchXpathEncodedThree.java:
```diff
@@ -0,0 +1,26 @@
+
+public class SuppressionXpathRegressionMatchXpathEncodedThree {
+
+    String newLineChar = "testFive\n" + // warning
```
I would add it another file. It makes easy to fix bugs, and add related test cases to '`r' later in time if required.

AkMo3 replied:
Done

AkMo3 commented:
> 
> 
> @AkMo3 there is one more thing I forgot about - there was same issue with escapes in char literals, e.g. `char a = '&'`
> Please extend test cases with this case (and other chars) as well.
> 
> Note - to produce violation, you can use IllegalTokenCheck with tokens = CHAR_LITERAL

Added to regression test. I will add the logs to this comment.

AkMo3 commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilterTest.java:
```diff
@@ -179,4 +182,17 @@ private static SuppressionXpathFilter createSuppressionXpathFilter(String fileNa
     }
 
+    @Test
+    public void testAcceptThree() throws Exception {
```
@strkkk Sir, can you please explain how this check work. On whatever input I give, it always asserts true which should not be valid.

AkMo3 replied:
I got it. Added tests for both String Literal and Char literal.

AkMo3 reviewed (commented):
@strkkk How is TreeWalker audit test accepting FileContent `null` in [SuppressionXpathFilterTest.java#L145](https://github.com/checkstyle/checkstyle/blob/7c7958a6004f3a0bd4c9444027ecdffc8ae9353f/src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilterTest.java#L145) . The first parameter is FileContent, and second is `"file1.java"` is just doesn't exist. How is this check even working?

AkMo3 commented:
@strkkk Can you please tell why wrecker failed?

strkkk commented:
> Can you please tell why wrecker failed?

Not related to your code, there is common stackoverflow error

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,61 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
     }
+
+    /**
+     * Controls true escape characters. True escape characters need '&' before but it also
```
`True escape characters need '&' before but it also`
here & should be escaped as well

AkMo3 replied:
On it

AkMo3 replied:
Done

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionxpathfilter/InputSuppressionXpathFilterEscape.xml:
```diff
@@ -0,0 +1,54 @@
+    <suppress-xpath
+            files="Test"
+            id="777"
+            query="/CLASS_DEF[./IDENT[@text='InputSuppressionXpathFilterEscape']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='quoteChar']]/ASSIGN/EXPR/PLUS[./STRING_LITERAL[@text='\&quot;escaped\&quot;']]"/>
```
please wrap this long lines like it is done in 
https://github.com/checkstyle/checkstyle/blob/master/config/suppressions-xpath.xml#L9

AkMo3 replied:
Thanks for the reference. On it.

AkMo3 replied:
Done

strkkk commented on config/checkstyle_non_main_files_suppressions.xml:
```diff
@@ -87,6 +87,8 @@
   <suppress id="lineLength"
          files="src[\\/]test[\\/]resources[\\/].*[\\/]customimportorder"/>
+  <suppress id="lineLength"
+         files="src[\\/]test[\\/]resources[\\/].*[\\/]filters[\\/]suppressionxpathfilter[\\/]InputSuppressionXpathFilterEscape.xml"/>
```
please remove this suppression and wrap lines in file

AkMo3 replied:
Ok. I will keep that in mind.

AkMo3 replied:
Done

strkkk commented on config/checkstyle_input_suppressions.xml:
```diff
@@ -2659,6 +2661,8 @@
   <suppress id="StyleValidationCommentInInputs"
             files="filters[\\/]suppressionxpathfilter[\\/]InputSuppressionXpathFilter.java"/>
+  <suppress id="StyleValidationCommentInInputs"
+            files="filters[\\/]suppressionxpathfilter[\\/]InputSuppressionXpathFilterEscape.java"/>
```
Please remove this suppressions.
Add config comment to input file (it is even better to split input file to 2 files for strings and chars)

AkMo3 replied:
On it.

AkMo3 replied:
Done

AkMo3 commented:
> Not related to your code, there is common stackoverflow error

Ok, thanks. Sir, can you please review again. I had done all the changes you suggested. Also, why are normal contributors not given permission to see wrecker result?

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionxpathfilter/InputSuppressionXpathFilterEscape.java:
```diff
@@ -0,0 +1,35 @@
+
+    char d = '<'; // warning
+
+    char e = '>'; // warning
```
`// warning` -> // violation

Please move char examples to other input file

AkMo3 replied:
Ok, On it.

AkMo3 replied:
Sir, can you please tell how to solve `StyleValidationCommentInInputs` error on this file
```
package com.puppycrawl.tools.checkstyle.filters.suppressionxpathfilter;

/*
 * Config: Default
 */
public class InputSuppressionXpathFilterEscapeChar {

    char a = '&'; // violation

    char b = '\"'; // violation

    char c = '\''; // violation

    char d = '<'; // violation

    char e = '>'; // violation
}
```

AkMo3 replied:
I am struggling to solve this error for more than 2 hrs, but unable to pass this check. Also, looked at regexp of this check, it matches (according to me) to the [checkstyle_input_checks.xml](https://github.com/checkstyle/checkstyle/blob/7c7958a6004f3a0bd4c9444027ecdffc8ae9353f/config/checkstyle_input_checks.xml#L29) but still gives this error.

strkkk replied:
it is weird. Please commit your changes and push them, I can checkout and take a look

AkMo3 replied:
I had pushed changes Sir. Please take a look.

AkMo3 replied:
After commiting the issue seems to be solved automatically. Sorry for bothering you Sir.

AkMo3 replied:
Changes done.

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilterTest.java:
```diff
@@ -179,4 +179,108 @@ private static SuppressionXpathFilter createSuppressionXpathFilter(String fileNa
+                "", "", null, null, "777", getClass(), null);
+        final TreeWalkerAuditEvent evApos = new TreeWalkerAuditEvent(null,
+                "Test.java", messageApos, JavaParser.parseFile(file,
+                JavaParser.Options.WITHOUT_COMMENTS));
```
```
final LocalizedMessage messageApos = new LocalizedMessage(23, 34, TokenTypes.PLUS,
                "", "", null, null, "777", getClass(), null);
        final TreeWalkerAuditEvent evApos = new TreeWalkerAuditEvent(null,
                "Test.java", messageApos, JavaParser.parseFile(file,
                JavaParser.Options.WITHOUT_COMMENTS));
```
All these lines duplicated many times, only line number and column number are different. Please extract it to method that creates TreeWalkerAuditEvent

AkMo3 replied:
All right. On it.

AkMo3 replied:
Done

strkkk commented:
@AkMo3 Please reply on review comments when you want to discuss item or when you are done with it. 
The reason is pretty simple - each comment that do not follow these rules creates a redundant notification, which is not very pleasant for reviewers.

AkMo3 commented:
@strkkk Sir, please review.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionxpathfilter/InputSuppressionXpathFilterEscapeChar.java:
```diff
@@ -0,0 +1,18 @@
+    char d = '<';
+
+    char e = '>';
+    // violation
```
this violation comment looks weird. Please place comments correctly (to each line with violations)

AkMo3 replied:
Done.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionxpathfilter/InputSuppressionXpathFilterEscapeChar.java:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Config: Default.
+ */
+public class InputSuppressionXpathFilterEscapeChar { // violation
```
I doubt there is a violation :)

AkMo3 replied:
Yeah, sorry. I was trying all possibilities at that time. Done now.

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilterTest.java:
```diff
@@ -179,4 +180,85 @@ private static SuppressionXpathFilter createSuppressionXpathFilter(String fileNa
+    private LocalizedMessage getMessageForCharLiteral(int lineNo, int columnNo) {
+        return new LocalizedMessage(lineNo, columnNo, TokenTypes.CHAR_LITERAL, "", "",
+                null, null, "777", getClass(), null);
+    }
```
it is better now, but can be improved even more. There is no need for separate methods to create a message. You can inline it into TreeWalkerAudit creation method.

Method signature will be like
```
private static TreeWalkerAuditEvent getTreeWalkerAuditForString(int lineNo, int columnNo, TokenTypes tokenType, File file)
```

AkMo3 replied:
Done

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilterTest.java:
```diff
@@ -179,4 +180,85 @@ private static SuppressionXpathFilter createSuppressionXpathFilter(String fileNa
+                null, null, "777", getClass(), null);
+    }
+
+    private static TreeWalkerAuditEvent getTreeWalkerAuditForString(LocalizedMessage messageQuot,
```
`getTreeWalkerAuditForString` -> createTreeWalkerAuditEvent

AkMo3 replied:
Done

AkMo3 commented:
GitHub, rebase

AkMo3 commented:
@strkkk Changes done. Please review.

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilterTest.java:
```diff
@@ -179,4 +180,79 @@ private static SuppressionXpathFilter createSuppressionXpathFilter(String fileNa
+            message = getMessageForCharLiteral(lineNo, columnNo);
+        }
+        else {
+            message = getMessageForStringLiteral(lineNo, columnNo);
```
no need for methods and if-else block here. Just
```
final LocalizedMessage message = new LocalizedMessage(lineNo, columnNo, tokenTypes, "", "",
                null, null, "777", SuppressionXpathFilterTest.class, null);
```
is enough.
`getMessageFor...` methods can be removed

AkMo3 replied:
Done

strkkk replied:
Not done. There is a method left `getMessageForStringLiteral`, it is not needed. Please inline it.

AkMo3 replied:
Sorry, I misinterpreted it. It`s done now. Also, I had removed the method, but I had not totally inlined it inside TreeWalkerAuditEvenet as it looks cluttered a lot. So, rather initialised it inside a message variable and then used that.

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilterTest.java:
```diff
@@ -179,4 +180,79 @@ private static SuppressionXpathFilter createSuppressionXpathFilter(String fileNa
+                null, null, "777", SuppressionXpathFilterTest.class, null);
+    }
+
+    private static TreeWalkerAuditEvent createTreeWalkerAudit(int lineNo, int columnNo,
```
please move this method to bottom of the class, after all test methods. It is a common thing to have test cases at the top and all tech methods at the bottom.

AkMo3 replied:
Done

strkkk reviewed (changes requested):
Few last minors

pbludov commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -458,4 +460,28 @@ public void testConstructorWithTreeWalkerAuditEvent() {
+        encoding.put("\n", "&#10;");
+        encoding.put("\r", "");
+
+        for (Map.Entry<String, String> put : encoding.entrySet()) {
```
Why is this variable named `put`? Maybe this is the `entry` or something like that?

AkMo3 replied:
Done

pbludov commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -458,4 +460,28 @@ public void testConstructorWithTreeWalkerAuditEvent() {
+
+        for (Map.Entry<String, String> put : encoding.entrySet()) {
+            final String encoded = XpathQueryGenerator.encode(put.getKey());
+            assertEquals(put.getValue(), encoded, "\"" + put.getKey() + "\"");
```
Please use `truth.assertWithMessage("...").that(...)` here.

AkMo3 replied:
@pbludov I changed `assertEquals(put.getValue(), encoded, "\"" + put.getKey() + "\"")` -> `Truth.assertWithMessage("Incorrect encoding").that(entry.getValue()).isEqualTo(encoded)` but it is giving me pmd error of ` Rule:JUnitTestsShouldIncludeAssert Priority:3 JUnit tests should include assert() or fail().` So I had to include both assertEquals and Truth test test. Is it Ok?

pbludov replied:
Just use static import. See this [example](https://github.com/checkstyle/checkstyle/blob/1e3fb3fae940bf7ea340592f7f0cda3af1b320e3/src/test/java/com/puppycrawl/tools/checkstyle/utils/XpathUtilTest.java#L178)

AkMo3 replied:
Done

pbludov reviewed (changes requested):
Please do two minor changes

romani commented on src/it/resources/org/checkstyle/suppressionxpathfilter/matchxpath/SuppressionXpathRegressionMatchXpathEncodedAmpChar.java:
```diff
@@ -0,0 +1,25 @@
+////////////////////////////////////////////////////////////////////////////////
```
please remove header from all test Inputs files, such files should contain only content that is matter for test.

AkMo3 replied:
Done

romani commented on src/it/resources/org/checkstyle/suppressionxpathfilter/matchxpath/SuppressionXpathRegressionMatchXpathEncodedFive.java:
```diff
@@ -0,0 +1,26 @@
+
+package org.checkstyle.suppressionxpathfilter.matchxpath;
+
+public class SuppressionXpathRegressionMatchXpathEncodedFive {
+
+    String ampersandChar = "&testThree" + // warning
```
please put in name of class target symbol you try to validate - "Amp".

please apply this rule to all other files that have in name "One, Two, Three, .... Five". Lets have names self descriptive

AkMo3 replied:
Done

AkMo3 replied:
Even I don't know but I didn't change it since it was mentioned in [XMLLogger](https://github.com/checkstyle/checkstyle/blob/5f53bfe41e96a0fcd0a8266ffc0db0fb32c54d6e/src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java#L275) for the exact same thing, so I didn't change the description. I had changed it now.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,61 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+     * @param chr character to check.
+     * @param encStr StringBuilder from parent method encode().
+     */
+    private static void controlTrueCharacter(char chr, StringBuilder encStr) {
```
please move this method to be after "encode" method, we care about order of method declarations.

romani replied:
name of this method is weird.
What is "true escape character" ?

Can you inline method in switch ? to make sure we see all `append` in same place

AkMo3 replied:
Even I don't know but I didn't change it since it was mentioned in [XMLLogger](https://github.com/checkstyle/checkstyle/blob/5f53bfe41e96a0fcd0a8266ffc0db0fb32c54d6e/src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java#L275) for the exact same thing, so I didn't change the description. I had changed it now.

AkMo3 replied:
I didn't get by **inline method in switch**. Can you please elaborate a bit.

romani replied:
Move whole body of this method to switch code block. Remove this method at all

AkMo3 replied:
Sir, the issue with it is the `encode` method is becoming larger than 30 lines, and then it gives violation.

romani replied:
please read https://github.com/checkstyle/checkstyle/pull/9247#discussion_r576457648

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,61 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+    public static String encode(String value) {
+        final StringBuilder sb = new StringBuilder(256);
+        for (int i = 0; i < value.length(); i++) {
+            final char chr = value.charAt(i);
```
we do not use "i"
can we use `value.chars().forEach` ?

AkMo3 replied:
I used the regular for each for string, I hope that might be ok.

AkMo3 replied:
Sir, this is also resolved according to me, as I have replaced the normal for against a foreach loop.

romani replied:
Your approach is using copy of memory, 
Steams approach does not make copy 
https://github.com/bpupadhyaya/openjdk-8/blob/45af329463a45955ea2759b89cb0ebfe40570c3f/jdk/src/share/classes/java/lang/CharSequence.java#L130

romani replied:
not done, but https://github.com/checkstyle/checkstyle/pull/9247#discussion_r576457648 looks like a way out.

AkMo3 replied:
I had used value.chars().forEach, but there is a check which prevents for than 10 lines for `Lambda Expression` so I had to make a new method and then use it. Please tell if it works or nor?

romani reviewed (changes requested):
items:

romani commented on src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionMatchXpathTest.java:
```diff
@@ -87,4 +89,322 @@ public void testTwo() throws Exception {
+    }
+
+    @Test
+    public void testEncodedTwo() throws Exception {
```
Fix names of method, same as you did with class name

AkMo3 replied:
Done

romani reviewed (approved):
Item:

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -75,23 +75,35 @@
-    /** The root ast. */
+    /**
+     * The root ast.
+     */
```
Please revert all unrelated updates of code, you can send them in separate PR

AkMo3 replied:
Done

romani reviewed (approved):
Items:

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,60 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+     * @param value the value to escape.
+     * @return the escaped value if necessary.
+     */
+    public static String encode(String value) {
```
I try to think why we need our special implementation of this 
can you reuse Guava - https://stackoverflow.com/a/52364912/1015848 ?

AkMo3 replied:
Sir, we need our special implementation because of two reasons: 
1. We use Saxon for decoding xpath, which uses some custom rules like double apos to represent single apos.
2. I am not sure if non-ASCII character required to be encoded, but we are doing it right now. But guava is not able to encode them till now as there documentation states `Currently the escapers provided by this class do not escape any characters outside the ASCII character range`.

romani replied:
Please put such reasons to javadoc of method, it is not obvious, let's keep this knowledge written

AkMo3 replied:
Done.

AkMo3 replied:
Also, please see [#9247 Comment](https://github.com/checkstyle/checkstyle/pull/9247#discussion_r576550029) .

romani replied:
yes all details to javadoc.

romani reviewed (changes requested):
item:

AkMo3 commented:
GitHub, rebase

AkMo3 commented:
@romani Sir I feel that value.chars().forEach() approach may not work since it returns an intStream (ASCII value of char) but our requirement is for non-ASCII values too. Also, TeamCity gives this error `Cast from 'int' to char may result in loss of precision `. Please share your views.

romani commented on src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionMatchXpathTest.java:
```diff
@@ -87,4 +89,322 @@ public void testTwo() throws Exception {
+            new File(getPath("SuppressionXpathRegressionMatchXpathEncodedCarriageString.java"));
+
+        final DefaultConfiguration moduleConfig =
+                createModuleConfig(OperatorWrapCheck.class);
```
please reuse your idea on `moduleConfig.addAttribute("tokens", "CHAR_LITERAL");`
and update String based test with usage of the same Check - IllegalTokenCheck
https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#STRING_LITERAL

AkMo3 replied:
Done

romani commented on src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionMatchXpathTest.java:
```diff
@@ -87,4 +89,322 @@ public void testTwo() throws Exception {
+        final DefaultConfiguration moduleConfig =
+                createModuleConfig(IllegalTokenCheck.class);
+
+        moduleConfig.addAttribute("tokens", "CHAR_LITERAL");
```
please remove empty line here and in all other similar cases that you add, it is same block of `moduleConfig` creation.

AkMo3 replied:
Done

romani reviewed (changes requested):
items:

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,74 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+     * @param encStr StringBuilder from parent method encode().
+     * @param chr Character to check.
+     */
+    private static void encodeCharacter(StringBuilder encStr, char chr) {
```
make this method to return `char` to let upper method that has StringBuilder do append.

AkMo3 replied:
Done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,74 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+                break;
+            default:
+                controlCharacter(chr, encStr);
+                break;
```
make it like:
```
            case '\r':
                break;
            case '\n':
                encStr.append("&#10;");
                break;
            default:
                if (Character.isISOControl(chr)) {
                    encStr.append("#x").append(Integer.toHexString(chr)).append(';');
                }
                else {
                    encStr.append(chr);
                }
                break;
        }
    }
```

AkMo3 replied:
Done.

AkMo3 replied:
Sir, this is also resolved I think.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,74 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+     */
+    public static String encode(String value) {
+        final StringBuilder sb = new StringBuilder(256);
+        value.chars().forEach(
```
instead of chars please use "codePoints()" to work with unicode.
Please make Input + test method that use unicode and it would fail if we use `chars()` here

AkMo3 replied:
Sir, can you please suggest how do I convert an Integer (ASCII value of char given by codepoints) to char without cast, as cast is not allowed by TeamCity.

romani replied:
you do `encodeCharacter(Character.toChars(chr)[0])` now .. why `[0]` ?

Please create Input file with String value like `"语言处理"` and how it can be escaped/suppressed by xpath.

AkMo3 replied:
`[0]` is used because `.toChars(chr)` returns a character array and we require only the first element of the array. Alright, I will add those examples.

romani replied:
please show by UT that encoding of unicode symbols continue to work fine.
plase add unicoded strings to  `public void testEncode() {`

AkMo3 replied:
Done.

romani replied:
please add unicode in string constants - it is allowed.
as very very last resort please add suppression for this file+check at https://github.com/checkstyle/checkstyle/blob/master/config/suppressions.xml

AkMo3 replied:
Done, No more required as they are not being encoded.

romani reviewed (changes requested):
items:

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -458,4 +461,29 @@ public void testConstructorWithTreeWalkerAuditEvent() {
+        encoding.put("\r", "");
+
+        for (Map.Entry<String, String> entry : encoding.entrySet()) {
+            final String encoded = XpathQueryGenerator.encode(entry.getKey());
```
please try to make this method (`encode`) as private and please test it through general execution of checkstyle validation `XpathRegressionMatchXpathTest.java`

AkMo3 replied:
I believe that this is also resolved since we are no longer implementing check`s similar to regression.

romani replied:
please do a test like testTabWidthAfterVoidLiteral
it will a bit verbose but actually same how it will work.
you can put all such small string to single or two String value and validate whole string `"&amp;lt;&amp;#0#x80;"`

AkMo3 replied:
Done

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionxpathfilter/InputSuppressionXpathFilterEscapeString.java:
```diff
@@ -0,0 +1,28 @@
+public class InputSuppressionXpathFilterEscapeString {
+
+    String quoteChar = "\"escaped\"" +
+            "2"; // violation
```
please use IllegalTokenCheck for this input to simplify input file.

AkMo3 replied:
Done

romani reviewed (changes requested):
items:

AkMo3 commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilterTest.java:
```diff
@@ -179,4 +180,64 @@ private static SuppressionXpathFilter createSuppressionXpathFilter(String fileNa
+        final File file = new File(getPath("InputSuppressionXpathFilterEscapeString.java"));
+
+        assertFalse(filter.accept(createTreeWalkerAudit(8, 37, TokenTypes.PLUS, file)),
+                "TreeWalker audit event should be rejected");
```
It would be great. Just a small issue, since the approach we take, will regression test is we check a file which only has a single violation, and a single Xpath is given for that suppression. That approach would lead to a lot file, with each file containing a single violation. So, how would we escape that. Please explain.

romani replied:
>  will regression test is we check a file which only has a single violation, and a single Xpath is given for that suppression

it could contain one of few. Execution without suppression files confirms amount of violations, execution with suppression file result in no violations.
Example - https://github.com/checkstyle/checkstyle/blob/4860e1005ea970fa5d177cc2640b1f34a3a92a96/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionMatchXpathTest.java#L42
please run this test in debug to see it in action.

AkMo3 replied:
Sir, will I was trying to implement this test, what I found was these tests extends to `AbstractXpathTestSupport` and our normal tests and supported by `AbstractPathTestSupport` and `AbstractModuleTestSupport` which don't have funtionality to process xpath and verify. 
The `AbstractXpathTestSupport` has root directory of `src/it` but we cannot use that. So, either we have add those functionality to our other interfaces or change the directory of `AbstractXpathTestSupport` and both of them sounds good to be.
Also, only one of them can be applied at a time (Keeping the class as `class` and not `interface`) so the entire test have to be changed to newer one. 
Please let me know the solutions to this problems or correct me if I am wrong in any of the above. Thanks.

AkMo3 replied:
@romani ping.

romani replied:
ok, lets keep it as is for now, we will resolve this scope of bigger project.

AkMo3 replied:
Sir, should I move ahead and create a new Issue for this, and then work on it to create the basic framework for this. I really like this idea and would like to work on it. Except semaphorica all other test were passed, and if I give some more time, I think I can really implement it, just some I require your and other members help a bit.

AkMo3 commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -458,4 +472,72 @@ public void testConstructorWithTreeWalkerAuditEvent() {
+    }
+
+    @Test
+    public void testEncodedGreaterChar() throws Exception {
```
@romani This is kind of test I have in mind. Please check if it works,

AkMo3 replied:
I had added both assertEquals also to pass pmd check.

AkMo3 reviewed (commented):
Please comment

AkMo3 reviewed (commented):
This test.

AkMo3 commented:
@romani Why is semaphoreci/pr giving me this error ?
```
1. ERROR in /home/semaphore/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java (at line 34) 00:24
	import org.checkstyle.suppressionxpathfilter.AbstractXpathTestSupport; 00:24
	       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 00:24
The import org.checkstyle.suppressionxpathfilter.AbstractXpathTestSupport cannot be resolved 00:24
---------- 00:24
2. ERROR in /home/semaphore/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java (at line 48) 00:24
	public class XpathQueryGeneratorTest extends AbstractXpathTestSupport { 00:24
	                                             ^^^^^^^^^^^^^^^^^^^^^^^^ 00:24
AbstractXpathTestSupport cannot be resolved to a type 00:24
---------- 00:24
3. ERROR in /home/semaphore/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java (at line 57) 00:24
	protected String getCheckName() { 00:24
	                 ^^^^^^^^^^^^^^ 00:24
The method getCheckName() of type XpathQueryGeneratorTest must override or implement a supertype method 00:24
---------- 00:24
4. ERROR in /home/semaphore/checkstyle/src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java (at line 62) 00:24
	protected String getPackageLocation() { 00:24
	                 ^^^^^^^^^^^^^^^^^^^^ 00:24
The method getPackageLocation() of type XpathQueryGeneratorTest must override or implement a supertype method
```

AkMo3 commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -458,4 +463,69 @@ public void testConstructorWithTreeWalkerAuditEvent() {
+    }
+
+    @Test
+    public void testEscapeCharacters() throws Exception {
```
This test replaces Earlier Escape Test.

AkMo3 commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -458,4 +463,69 @@ public void testConstructorWithTreeWalkerAuditEvent() {
     }
 
+    @Test
+    public void testUnicodeCharacters() throws Exception {
```
New Unicode Test.

AkMo3 commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -69,9 +72,9 @@ public void testClassDef() {
-            "/CLASS_DEF[./IDENT[@text='InputXpathQueryGenerator']]",
-            "/CLASS_DEF[./IDENT[@text='InputXpathQueryGenerator']]/MODIFIERS",
-            "/CLASS_DEF[./IDENT[@text='InputXpathQueryGenerator']]/MODIFIERS/LITERAL_PUBLIC");
+                "/CLASS_DEF[./IDENT[@text='InputXpathQueryGenerator']]",
```
Indention will be fixed in next commit.

AkMo3 commented on src/test/resources/com/puppycrawl/tools/checkstyle/xpath/xpathquerygenerator/InputXpathQueryGeneratorUnicodeCharacters.java:
```diff
@@ -0,0 +1,5 @@
+package com.puppycrawl.tools.checkstyle.xpath.xpathquerygenerator;
+
+public class InputXpathQueryGeneratorUnicodeCharacters {
+    String unicode = "Ǆǚǯǣ";
```
Are more Unicode Char needed?

AkMo3 reviewed (commented):
Some points:

AkMo3 commented on src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java:
```diff
@@ -359,4 +359,19 @@ protected String getResourceLocation() {
         return "it";
     }
 
+    /**
```
Done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java:
```diff
@@ -332,4 +332,75 @@ private boolean isMatchingByLineAndColumnAndTokenType(DetailAST ast) {
+     * @param value the value to escape.
+     * @return the escaped value if necessary.
+     */
+    public static String encode(String value) {
```
We have raletively easy way to test this method by input files.
Please make it private.

AkMo3 replied:
Sir, I will not be able to do changes till the 8th of march due to my examinations. So, please wait till then. Also, please answer this [#9247 (comment)](https://github.com/checkstyle/checkstyle/pull/9247#discussion_r582811838). It would be helpful.

AkMo3 replied:
Done

romani reviewed (changes requested):
Item:

AkMo3 commented:
@romani What these new tests show that unicode characters and some other char like `\n` and `\r` are not required to be encoded and are handled itself saxon. So I had removed them from X-Path Query generator and removed the unicode test as well from `XpathQueryGeneratorTest`.

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -196,7 +196,7 @@ public void testImport() {
         final int columnNumber = 1;
         final XpathQueryGenerator queryGenerator = new XpathQueryGenerator(rootAst, lineNumber,
-                columnNumber, fileText, DEFAULT_TAB_WIDTH);
+            columnNumber, fileText, DEFAULT_TAB_WIDTH);
```
Please restore indentation

AkMo3 replied:
Done

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/xpath/xpathquerygenerator/InputXpathQueryGeneratorEscapeCharacters.java:
```diff
@@ -0,0 +1,7 @@
+public class InputXpathQueryGeneratorEscapeCharacters {
+    String testOne = "<>'\"&abc;&lt;\u0080\n";
+
+    String testTwo = "&#0;&#X0\u0001\n";
```
Please add test case with text block
New jdk feature where - multi line string.
Better to create new Input file for such sofisticated input. It will have invisible \n

AkMo3 replied:
Done, but `TEXT_BLOCK_CONTENT` are not suppressed like String, the entire block is suppressed not line and the inside content doesn't matter if there is any violation inside block.

romani reviewed (changes requested):
Items:

AkMo3 commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/xpath/xpathquerygenerator/InputXpathQueryGeneratorTextBlock.java:
```diff
@@ -0,0 +1,10 @@
+package com.puppycrawl.tools.checkstyle.xpath.xpathquerygenerator;
+
+public class InputXpathQueryGeneratorTextBlock {
+
+    String testOne = """               // violation
+        &This is the first line
+        >This is the second line
+        <This is the third line
+        """;
+}
```
Drone PR failing since **TextBlock** are Supported from SE 15 and non-compilable with Java 8. How to escape this test?

AkMo3 replied:
Fixed.

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGeneratorTest.java:
```diff
@@ -458,4 +460,71 @@ public void testConstructorWithTreeWalkerAuditEvent() {
+        final List<String> expectedTestOne = Collections.singletonList(
+            "/CLASS_DEF[./IDENT[@text='InputXpathQueryGeneratorTextBlock']]/OBJBLOCK/"
+                    + "VARIABLE_DEF[./IDENT[@text='testOne']]/ASSIGN/EXPR/"
+                    + "TEXT_BLOCK_LITERAL_BEGIN/TEXT_BLOCK_CONTENT"
```
do we have ability to check escaped content of text block here  like `TEXT_BLOCK_CONTENT[@text='.......']`?

AkMo3 replied:
No, I ran IlleagalTokenCheck and the suppressions generated doesn't contain any text attribute.

romani replied:
> suppressions generated

please create Xpath manually.

AkMo3 replied:
If by manual generation you mean by using [XpathQueryGenerator.generate()](https://github.com/checkstyle/checkstyle/blob/2275613f2991e6e5df78e629c3ba379c9bcce6e9/src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java#L142) then that too doesn't contain text attributes. I was also sceptical about the Xpath generated so had confirm by running a check.

AkMo3 replied:
> you can reduce text block to minimilistic to keep xpath in sane size

    String testOne = """
        &1line
       >2line
        <3line
        """;


Done

romani replied:
please refactor testTextBlocks to not use generator at all.
please make expectedTestOne to have `TEXT_BLOCK_CONTENT[@text='&1line\n>2line\n<3line']` and prove by UT that such xpath suppress violation.

romani replied:
moved to https://github.com/checkstyle/checkstyle/issues/9359

romani reviewed (changes requested):
last item:

AkMo3 commented:
Test File
```
//non-compiled with javac: Compilable with Java14
package com.puppycrawl.tools.checkstyle.xpath.xpathquerygenerator;

public class InputXpathQueryGeneratorTextBlock {

    String testOne = """
        &1line
        >2line
        <3line
        """;
}
```

Config.xml 
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="IllegalToken">
      <property name="tokens" value="STRING_LITERAL"/>
      <property name="tokens" value="TEXT_BLOCK_CONTENT"/>
    </module>
  </module>
</module>
```
Suppression generated with this PR
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suppressions PUBLIC
    "-//Checkstyle//DTD SuppressionXpathFilter Experimental Configuration 1.2//EN"
    "https://checkstyle.org/dtds/suppressions_1_2_xpath_experimental.dtd">
<suppressions>
<suppress-xpath
       files="InputXpathQueryGeneratorTextBlock.java"
       checks="IllegalTokenCheck"
       query="/CLASS_DEF[./IDENT[@text='InputXpathQueryGeneratorTextBlock']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='testOne']]/ASSIGN/EXPR/TEXT_BLOCK_LITERAL_BEGIN/TEXT_BLOCK_CONTENT"/>
</suppressions>
```
New Suppresion with `text` attribute
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suppressions PUBLIC
    "-//Checkstyle//DTD SuppressionXpathFilter Experimental Configuration 1.2//EN"
    "https://checkstyle.org/dtds/suppressions_1_2_xpath_experimental.dtd">
<suppressions>
<suppress-xpath
       files="InputXpathQueryGeneratorTextBlock.java"
       checks="IllegalTokenCheck"
       query="/CLASS_DEF[./IDENT[@text='InputXpathQueryGeneratorTextBlock']]/OBJBLOCK/VARIABLE_DEF[./IDENT[@text='testOne']]/ASSIGN/EXPR/TEXT_BLOCK_LITERAL_BEGIN/TEXT_BLOCK_CONTENT[@text='&amp;1line\n&gt;2line\n&lt;3line']"/>
</suppressions>
```
Result with normal (without text) 
```
Starting audit...
Audit done.
```

Result with text attribute
```
Starting audit...
[ERROR] /home/akmo/GitHub/Test/InputXpathQueryGeneratorTextBlock.java:6:25: Using '
        &1line
        >2line
        <3line
        ' is not allowed. [IllegalToken]
Audit done.
Checkstyle ends with 1 errors.
```

@romani I think we need to make a new Issue.

romani commented:
@AkMo3, please create new issue, looks like we have problem with match by content of text block

romani commented:
new issue is created - https://github.com/checkstyle/checkstyle/issues/9359

romani reviewed (approved):
ok to merge

@AkMo3 , thanks a lot for big update !!!!
